File: app/src/main/java/com/dn/sports/StepServices.java
package com.dn.sports;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.graphics.Color;
import android.os.Build;
import android.os.IBinder;
import android.widget.RemoteViews;

import androidx.annotation.Nullable;

import com.dn.sports.adcoinLogin.LoginListener;
import com.dn.sports.adcoinLogin.StepUserManager;
import com.dn.sports.utils.SharedPreferenceUtil;
import com.dn.sports.utils.Utils;

import java.text.NumberFormat;

/**
 * 显示通知栏服务，监听步数
 * */
public class StepServices extends Service {

    private StepUserManager stepUserManager = StepUserManager.getInstance();
    private NumberFormat nf;

    private LoginListener loginListener = new LoginListener(){
        @Override
        public void onStepChange() {
            super.onStepChange();
            if(remoteViews != null) {
                int steps = stepUserManager.getTodaySteps();
                float kmiles = ((float) steps) * 0.6f / 1000;
                String data = steps + "步" + " | ";
                data = data + nf.format(kmiles) + "公里" + " | ";
                float times = kmiles * 14;
                data = data + (int) ((times / 60) * 240) + "千卡";
                remoteViews.setTextViewText(R.id.notify_data, data);
                builder.setContent(remoteViews);
                startForeground(110, builder.build());
            }
        }
    };

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        StepUserManager.getInstance().setLoginListener(loginListener);
        nf = NumberFormat.getNumberInstance();
        nf.setMaximumFractionDigits(2);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if(builder == null || remoteViews == null) {
            addNotify();
        }
        return START_STICKY;
    }

    private Notification.Builder builder;
    private RemoteViews remoteViews;
    public static final String START_FORM_SERVICE = "start_form_service";
    public final static String INTENT_BUTTONID_TAG = "ButtonId";
    public final static int BUTTON_OPEN_APP = 1;
    public final String CHANNEL_ONE_ID = "com.steps";
    public final String CHANNEL_ONE_NAME = "ChannelSteps";

    public void addNotify(){

        boolean open = (boolean) SharedPreferenceUtil.Companion.getInstance(this).get("testFeedMessage", true);
        if (!open) {
            return;
        }

        builder = new Notification.Builder
                (this.getApplicationContext());
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            NotificationChannel notificationChannel;
            notificationChannel = new NotificationChannel(CHANNEL_ONE_ID,
                    CHANNEL_ONE_NAME, NotificationManager.IMPORTANCE_HIGH);
            notificationChannel.enableLights(true);
            notificationChannel.setImportance(NotificationManager.IMPORTANCE_LOW);
            notificationChannel.setLightColor(Color.RED);
            notificationChannel.setShowBadge(true);
            notificationChannel.setLockscreenVisibility(Notification.VISIBILITY_SECRET);
            NotificationManager manager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
            manager.createNotificationChannel(notificationChannel);
            builder.setChannelId(CHANNEL_ONE_ID);
        }
        remoteViews = new RemoteViews(getPackageName(), R.layout.notify_layout);//通知栏布局
        Intent openIt = new Intent(this, MainActivity.class);
        openIt.putExtra(START_FORM_SERVICE,true);
        openIt.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        openIt.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
        PendingIntent itP = PendingIntent.getActivity(this, BUTTON_OPEN_APP, openIt, PendingIntent.FLAG_UPDATE_CURRENT);
        remoteViews.setOnClickPendingIntent(R.id.notify_root_layout, itP);

        remoteViews.setTextViewText(R.id.notify_image_title, Utils.getTopTitleName(this));
        builder.setSmallIcon(R.mipmap.ic_app);
        builder.setContent(remoteViews);
        builder.setSound(null,null);
        if(Build.VERSION.SDK_INT>=21)
            builder.setVisibility(Notification.VISIBILITY_SECRET);

        startForeground(110, builder.build());
    }
}

File: app/src/main/java/com/dn/sports/StepServices.kt
package com.dn.sports

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Build
import android.os.IBinder
import android.widget.RemoteViews
import androidx.core.app.NotificationCompat
import androidx.lifecycle.LifecycleService
import androidx.lifecycle.lifecycleScope
import com.dn.sports.adcoinLogin.LoginListener
import com.dn.sports.adcoinLogin.StepUserManager
import com.dn.sports.utils.SharedPreferenceUtil
import com.dn.sports.utils.Utils
import kotlinx.coroutines.launch
import java.text.NumberFormat

/**
 * Modernized Step Counting Service (Foreground Service 2.0)
 * Handles background step counting, notification updates, and data sync to Room.
 */
class StepServices : LifecycleService() {

    private val stepUserManager = StepUserManager.getInstance()
    private val nf: NumberFormat by lazy {
        NumberFormat.getNumberInstance().apply {
            maximumFractionDigits = 2
        }
    }

    private var builder: NotificationCompat.Builder? = null
    private var remoteViews: RemoteViews? = null

    private val loginListener = object : com.dn.sports.adcoinLogin.LoginListener() {
        override fun onStepChange() {
            super.onStepChange()
            updateNotification()
            syncStepsToDatabase()
            updateWidget()
        }
    }

    private fun updateWidget() {
        // Persist today's steps for the Widget
        val steps = stepUserManager.todaySteps
        SharedPreferenceUtil.getInstance(this).put("today_steps", steps)

        // Notify Widget
        sendBroadcast(Intent("com.dn.sports.ACTION_UPDATE_WIDGET").apply {
            setPackage(packageName)
        })
    }

    override fun onCreate() {
        super.onCreate()
        stepUserManager.setLoginListener(loginListener)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        super.onStartCommand(intent, flags, startId)
        if (builder == null || remoteViews == null) {
            addNotify()
        }
        return START_STICKY
    }

    private fun addNotify() {
        val open = SharedPreferenceUtil.getInstance(this).get("testFeedMessage", false) as Boolean
        if (!open) return

        val channelId = "com.steps"
        val channelName = "ChannelSteps"

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            val channel = NotificationChannel(channelId, channelName, NotificationManager.IMPORTANCE_LOW).apply {
                enableLights(true)
                lightColor = Color.RED
                setShowBadge(true)
                lockscreenVisibility = Notification.VISIBILITY_SECRET
            }
            manager.createNotificationChannel(channel)
        }

        remoteViews = RemoteViews(packageName, R.layout.notify_layout).apply {
            val openIt = Intent(this@StepServices, MainActivity::class.java).apply {
                putExtra(START_FORM_SERVICE, true)
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_SINGLE_TOP)
            }
            val pendingIntent = PendingIntent.getActivity(
                this@StepServices, 1, openIt,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )
            setOnClickPendingIntent(R.id.notify_root_layout, pendingIntent)
            setTextViewText(R.id.notify_image_title, Utils.getTopTitleName(this@StepServices))
        }

        builder = NotificationCompat.Builder(this, channelId).apply {
            setSmallIcon(R.mipmap.ic_app)
            setContent(remoteViews)
            setSound(null)
            setOngoing(true)
            priority = NotificationCompat.PRIORITY_LOW
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                setVisibility(NotificationCompat.VISIBILITY_SECRET)
            }
        }

        updateNotification()
        startForeground(NOTIFICATION_ID, builder?.build())
    }

    private fun updateNotification() {
        remoteViews?.let { views ->
            val steps = stepUserManager.todaySteps
            val kmiles = steps * 0.6f / 1000
            val data = "${steps}步 | ${nf.format(kmiles)}公里 | ${(kmiles * 14 / 60 * 240).toInt()}千卡"

            views.setTextViewText(R.id.notify_data, data)
            builder?.setContent(views)

            val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.notify(NOTIFICATION_ID, builder?.build())
        }
    }

    private fun syncStepsToDatabase() {
        lifecycleScope.launch {
            val steps = stepUserManager.todaySteps
            StepApplication.getInstance().repository.upsertDailyStepsRecord(steps)
        }
    }

    override fun onDestroy() {
        stepUserManager.removeLoginListener(loginListener)
        super.onDestroy()
    }

    companion object {
        private const val NOTIFICATION_ID = 110
        const val START_FORM_SERVICE = "start_form_service"
    }
}
